File: src/puppy_core/puppy_core/battery_status_adapter.py
"""Adapt raw BatteryState into semantic BatteryStatus for puppy_core."""
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import BatteryState

from puppy_interfaces.msg import BatteryStatus


class BatteryStatusAdapterNode(Node):
    """Convert sensor_msgs/BatteryState into puppy_interfaces/BatteryStatus."""

    def __init__(self):
        super().__init__('battery_status_adapter')

        self.declare_parameter('low_battery_threshold', 0.20)
        self.declare_parameter('critical_battery_threshold', 0.10)
        self.low_battery_threshold = float(
            self.get_parameter('low_battery_threshold').value
        )
        self.critical_battery_threshold = float(
            self.get_parameter('critical_battery_threshold').value
        )

        self.pub = self.create_publisher(BatteryStatus, '/battery_status', 10)
        self.create_subscription(BatteryState, '/battery_state', self._on_battery, 10)

        self.get_logger().info('Battery status adapter started')

    def _on_battery(self, msg: BatteryState):
        status = BatteryStatus()
        status.header = msg.header
        status.voltage = float(msg.voltage)
        status.current = float(msg.current)
        status.percent = float(max(0.0, min(1.0, msg.percentage)))
        status.charging = msg.power_supply_status in (
            BatteryState.POWER_SUPPLY_STATUS_CHARGING,
            BatteryState.POWER_SUPPLY_STATUS_FULL,
        )
        status.low_battery = status.percent <= self.low_battery_threshold
        status.critical_battery = status.percent <= self.critical_battery_threshold
        self.pub.publish(status)


def main(args=None):
    rclpy.init(args=args)
    node = BatteryStatusAdapterNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()

File: cpp_src/puppy_core/battery_status_adapter.cpp
// Battery Status Adapter 实现
#include "puppy_core/battery_status_adapter.h"

#include <algorithm>

namespace puppy_core {

BatteryStatusAdapterNode::BatteryStatusAdapterNode()
    : rclcpp::Node("battery_status_adapter") {
  this->declare_parameter("low_battery_threshold", 0.20);
  this->declare_parameter("critical_battery_threshold", 0.10);
  low_battery_threshold_ =
      this->get_parameter("low_battery_threshold").as_double();
  critical_battery_threshold_ =
      this->get_parameter("critical_battery_threshold").as_double();

  pub_ = this->create_publisher<puppy_interfaces::msg::BatteryStatus>(
      "/battery_status", 10);
  sub_ = this->create_subscription<sensor_msgs::msg::BatteryState>(
      "/battery_state", 10,
      std::bind(&BatteryStatusAdapterNode::onBattery, this,
                std::placeholders::_1));

  RCLCPP_INFO(this->get_logger(), "Battery status adapter started");
}

void BatteryStatusAdapterNode::onBattery(
    const sensor_msgs::msg::BatteryState::SharedPtr msg) {
  puppy_interfaces::msg::BatteryStatus status;
  status.header = msg->header;
  status.voltage = msg->voltage;
  status.current = msg->current;
  status.percent = static_cast<float>(
      std::max(0.0, std::min(1.0, static_cast<double>(msg->percentage))));
  // Python 版使用 power_supply_status in (CHARGING, FULL)
  using BS = sensor_msgs::msg::BatteryState;
  status.charging = (msg->power_supply_status == BS::POWER_SUPPLY_STATUS_CHARGING) ||
                    (msg->power_supply_status == BS::POWER_SUPPLY_STATUS_FULL);
  status.low_battery = static_cast<double>(status.percent) <= low_battery_threshold_;
  status.critical_battery =
      static_cast<double>(status.percent) <= critical_battery_threshold_;
  pub_->publish(status);
}

}  // namespace puppy_core
